src/windows.rs

```diff
@@ -0,0 +1,464 @@
+                device_name: self.device_name.clone(),
+                function_name: "IOCTL_VIDEO_QUERY_SUPPORTED_BRIGHTNESS",
+                last_error: e,
+            }
+            .into()),
+        }
+    }
+}
+
+pub fn brightness_devices() -> impl Stream<Item = Result<Brightness, SysError>> {
+    unsafe extern "system" fn enum_monitors(
+        arg1: HMONITOR,
+        _arg2: HDC,
+        _arg3: *mut RECT,
+        arg4: LPARAM,
+    ) -> BOOL {
+        let monitors = &mut *(arg4.0 as *mut Vec<HMONITOR>);
+        monitors.push(arg1);
+        return true.into();
+    }
+    let mut hmonitors = Vec::<HMONITOR>::new();
+    let devices = unsafe {
+        if !EnumDisplayMonitors(
+            HDC::NULL,
+            std::ptr::null_mut() as *mut RECT,
+            Some(enum_monitors),
+            LPARAM(&mut hmonitors as *mut _ as isize),
+        )
+        .as_bool()
+        {
+            either::Left(iter::once(Err(SysError::ListingDevicesFailed(
+                "EnumDisplayMonitors",
+                GetLastError(),
+            ))))
+        } else {
+            either::Right(hmonitors.into_iter().flat_map(|hmonitor| {
+                // Get the Name of the HMONITOR
+                let mut info: MONITORINFOEXW = std::mem::MaybeUninit::zeroed().assume_init();
+                info.__AnonymousBase_winuser_L13558_C43.cbSize =
+                    std::mem::size_of::<MONITORINFOEXW>() as u32;
+                let info_ptr = &mut info as *mut _;
+                if !GetMonitorInfoW(hmonitor, info_ptr as *mut MONITORINFO).as_bool() {
+                    return vec![Err(SysError::ListingDevicesFailed(
+                        "GetMonitorInfoW",
+                        GetLastError(),
+                    ))];
+                };
+
+                // Get the Physical Monitors in the HMONITOR
+                let mut physical_number: u32 = 0;
+                if !BOOL(GetNumberOfPhysicalMonitorsFromHMONITOR(
+                    hmonitor,
+                    &mut physical_number,
+                ))
+                .as_bool()
+                {
+                    return vec![Err(SysError::ListingDevicesFailed(
+                        "GetNumberOfPhysicalMonitorsFromHMONITOR",
+                        GetLastError(),
+                    ))];
+                };
+                let mut physical_monitors = Vec::with_capacity(physical_number as usize);
+                if !BOOL(GetPhysicalMonitorsFromHMONITOR(
+                    &hmonitor,
+                    physical_number,
+                    physical_monitors.as_mut_ptr(),
+                ))
+                .as_bool()
+                {
+                    return vec![Err(SysError::ListingDevicesFailed(
+                        "GetPhysicalMonitorsFromHMONITOR",
+                        GetLastError(),
+                    ))];
+                }
+                physical_monitors.set_len(physical_number as usize);
```

Comment:
What if a monitor disappears between the call to `GetNumberOfPhysicalMonitorsFromHMONITOR` and the call to `GetPhysicalMonitorsFromHMONITOR`? This code can lead to reading a memory location that has not been written to. There is no reason to resort to calling `set_len` here. Initializing `physical_monitors` as follows ensures soundness:

```rust
let physical_monitors = {
    let monitor = PHYSICAL_MONITOR {
        hPhysicalMonitor: HANDLE::NULL,
        szPhysicalMonitorDescription: [0; 128],
    };
    vec![monitor; physical_number]
};
```

Reply:
From this point on, the code is in charge of a set of handles to physical monitors that need to be closed eventually. If anything panics after this line (and before the code that wraps them in a type implementing `Drop`), the handles are leaked.

Reply:
Yeah I guess this depends on how much you trust the Windows API, it seems most likely they should Error if `dwPhysicalMonitorArraySize != output length`. But sure it makes sense to get rid of the unsafe set_len anyways

Reply:
I tested calling `GetNumberOfPhysicalMonitorsFromHMONITOR` with a size larger than my number of physical monitors and it does return an error, so this may not be a problem after all. Nevertheless erring on the side of caution does not hurt. This function reminded me of other Win32 functions that take a buffer and its size and may not fill it all, but they usually tell how many elements they write, which `GetNumberOfPhysicalMonitorsFromHMONITOR` cannot do, so it really makes sense for it to return an error if fewer items are written than requested.